### PR TITLE
Onnx support

### PR DESCRIPTION
**Pull request by dc-dc-dc**
## Proposed changes

Opening up a draft pr to get some initial feedback on the onnx support approach. Decided to utilize the onnx python package as it comes with an interface and test suite to simplify the addition of onnx. 

There are still ops missing support, open to any support from the community as I am unfamiliar with some of these ops. 

Adds onnx backend support closing #290 

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: python/mlx/onnx.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+import mlx.core as mx
+import numpy as np
+import onnx
+from onnx.helper import tensor_dtype_to_np_dtype
+
+onnx_ops = importlib.import_module("mlx.onnx_ops")
+
+
+class MlxBackend:
```
Comment (awni) [inline python/mlx/onnx.py:12]: Could you explain a bit how this class is intended to be used?

**Review by awni — COMMENTED** [review]
This looks really cool! Thanks for getting it started. I'm still trying to wrap my head around how it works, if you can show one or two simple examples that would be really helpful to understand the usage and put it all in context!

**Comment by dc-dc-dc** [comment]
@awni I created a simple example under `examples/python/onnx_example.py` (Doesn't run just yet as I am missing Conv Op)

**Comment by awni** [comment]
> For exporting we would need to trace the mlx model and then construct the onnx representation.

Right of course. We don't expose anything which does graph tracing in Python (or really even in C++) yet. Exposing that is not too difficult, but the challenge might be mapping our traced ops (which are Primitives) to the ops that ONNX can handle. In some cases the mapping might not make sense, but hopefully they are mostly pretty 1:1.

> Originally was planning on just creating the runtime to allow for running models in this pr. Can have a follow up pr for exporting.

Totally fine to do it step by step. I'm still wondering though, if this should live in MLX or elsewhere 🤔.

As you pointed out Jax has a standalone runtime, I can't quite figure out how the PyTorch runtime works. Do you know?

One benefit to keeping it outside MLX is we can pin the MLX Onnx runtime to a specific version of MLX which will make it easier to maintain.

**Comment by dc-dc-dc** [comment]
> We don't expose anything which does graph tracing

Yeah this would be a pre-req for adding the ability to export to onnx or other formats like hlo for example.

> I can't quite figure out how the PyTorch runtime works.

I believe that torch only supports exporting to onnx, but I did see a couple requests in their issues for the ability to import

I can see the view of having this as a standalone repo instead of being in the core lib as it just builds on top of mlx, would also separate any onnx specific issues from core mlx issues.

**Comment by dc-dc-dc** [comment]
Closing and reopening under the new repo

**Pull request closed without merging**